pkg/controllers/nodepool/readiness/controller.go

```diff
@@ -0,0 +1,121 @@
+
+	"k8s.io/apimachinery/pkg/api/equality"
+
+	"sigs.k8s.io/karpenter/pkg/cloudprovider"
+	"sigs.k8s.io/karpenter/pkg/operator/injection"
+
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"sigs.k8s.io/karpenter/pkg/apis/v1beta1"
+)
+
+// Controller for the resource
+type Controller struct {
+	kubeClient    client.Client
+	cloudProvider cloudprovider.CloudProvider
+}
+
+// NewController is a constructor
+func NewController(kubeClient client.Client, cloudProvider cloudprovider.CloudProvider) *Controller {
+	return &Controller{
+		kubeClient:    kubeClient,
+		cloudProvider: cloudProvider,
+	}
+}
+
+func (c *Controller) Reconcile(ctx context.Context, nodePool *v1beta1.NodePool) (reconcile.Result, error) {
+	ctx = injection.WithControllerName(ctx, "nodepool.status")
+	stored := nodePool.DeepCopy()
+	supportedNC := c.cloudProvider.GetSupportedNodeClasses()
+
+	if len(supportedNC) == 0 {
+		panic("no supported node classes found for the cloud provider")
+	}
+	nodeClass := supportedNC[0]
+	if err := c.kubeClient.Get(ctx, client.ObjectKey{Name: nodePool.Spec.Template.Spec.NodeClassRef.Name}, nodeClass); err != nil {
+		// If nodeClass is not found then we set status condition on nodePool as not ready
+		if errors.IsNotFound(err) {
+			log.FromContext(ctx).WithValues("NodeClass", nodePool.Spec.Template.Spec.NodeClassRef.Name).Info("nodeClass not found")
+			nodePool.StatusConditions().SetFalse(v1beta1.ConditionTypeNodeClassReady, "UnresolvedNodeClass", "Unable to resolve nodeClass")
+		}
+		if !equality.Semantic.DeepEqual(stored, nodePool) {
+			if e := c.kubeClient.Status().Update(ctx, nodePool); e != nil {
+				if errors.IsConflict(e) {
+					return reconcile.Result{Requeue: true}, nil
+				}
+				return reconcile.Result{}, client.IgnoreNotFound(e)
+			}
+		}
+		return reconcile.Result{}, client.IgnoreNotFound(err)
+	}
+	c.setReadyCondition(nodePool, nodeClass)
+	if !equality.Semantic.DeepEqual(stored, nodePool) {
+		if err := c.kubeClient.Status().Update(ctx, nodePool); err != nil {
+			if errors.IsConflict(err) {
+				return reconcile.Result{Requeue: true}, nil
+			}
+			return reconcile.Result{}, client.IgnoreNotFound(err)
+		}
+	}
+	return reconcile.Result{}, nil
+}
+
+func (c *Controller) setReadyCondition(nodePool *v1beta1.NodePool, nodeClass status.Object) {
+	ready := nodeClass.StatusConditions().Get(status.ConditionReady)
+	if ready.IsUnknown() {
+		nodePool.StatusConditions().SetFalse(v1beta1.ConditionTypeNodeClassReady, "NodeClassReadinessUnknown", "Node Class Readiness Unknown")
+	}
+	if ready.IsFalse() {
+		nodePool.StatusConditions().SetFalse(v1beta1.ConditionTypeNodeClassReady, ready.Reason, ready.Message)
+	}
+	if ready.IsTrue() {
+		nodePool.StatusConditions().SetTrue(v1beta1.ConditionTypeNodeClassReady)
+	}
```

Comment:
nit: consider doing 1) rather than 2) to portray that they're mutually exclusive
1)
```
if
elif
else 
```
2)
```
if 
if
if
```